File: formacOS/Sources/AnyClipCore/TXTCodec.swift
import Foundation

/// Minimal DNS TXT record codec (RFC 6763 §6): each entry is one length
/// byte followed by "key=value" bytes. Used to advertise/parse the same
/// properties the Python zeroconf beacon uses. Entries over 255 bytes are
/// skipped (ours are all tiny).
public enum TXTCodec {
    public static func encode(_ entries: [(String, String)]) -> Data {
        var out = Data()
        for (key, value) in entries {
            let raw = Data("\(key)=\(value)".utf8)
            guard raw.count <= 255 else { continue }
            out.append(UInt8(raw.count))
            out.append(raw)
        }
        return out
    }

    public static func decode(_ data: Data) -> [String: String] {
        var result: [String: String] = [:]
        var i = data.startIndex
        while i < data.endIndex {
            let len = Int(data[i])
            i = data.index(after: i)
            // A zero-length entry ends the scan (break, not continue): we only
            // decode records we or zeroconf encoded, where 0x00 means "empty
            // TXT record", never padding between entries.
            guard len > 0,
                  let end = data.index(i, offsetBy: len, limitedBy: data.endIndex)
            else { break }
            if let s = String(data: data[i..<end], encoding: .utf8),
               let eq = s.firstIndex(of: "=") {
                result[String(s[..<eq])] = String(s[s.index(after: eq)...])
            }
            i = end
        }
        return result
    }
}

File: forwindows/src/AnyClipCore/TxtCodec.cs
using System.Text;

namespace AnyClip.Core;

/// Minimal DNS TXT codec (RFC 6763 §6): [len byte]["key=value"]. Entries
/// over 255 bytes are skipped; a zero-length entry ends the scan (we only
/// decode records we or zeroconf encoded). Port of the formacOS TXTCodec.
public static class TxtCodec
{
    public static byte[] Encode(IEnumerable<(string Key, string Value)> entries)
    {
        var output = new List<byte>();
        foreach (var (key, value) in entries)
        {
            var raw = Encoding.UTF8.GetBytes($"{key}={value}");
            if (raw.Length > 255) continue;
            output.Add((byte)raw.Length);
            output.AddRange(raw);
        }
        return output.ToArray();
    }

    public static Dictionary<string, string> Decode(byte[] data)
    {
        var result = new Dictionary<string, string>();
        int i = 0;
        while (i < data.Length)
        {
            int len = data[i];
            i += 1;
            if (len == 0 || i + len > data.Length) break;
            var s = Encoding.UTF8.GetString(data, i, len);
            int eq = s.IndexOf('=');
            if (eq > 0) result[s[..eq]] = s[(eq + 1)..];
            i += len;
        }
        return result;
    }
}
